design/seed_vm.md

```diff
@@ -0,0 +1,190 @@
+No-Frills Introduction to Lua 5.1 VM
+Instructions](http://underpop.free.fr/l/lua/docs/a-no-frills-introduction-to-lua-5.1-vm-instructions.pdf).
+
+## SeedLang Value Types
+
+SeedLang is a dynamically typed programming language. Types are attached to
+values rather then variables. There are 9 types in SeedLang: Nil, Boolean,
+Number, String, List, Table, Function, Module and SeedObject:
+
+- Nil: Nil value, only one value nil
+- Boolean: Boolean value, true or false
+- Number: Double precision floating point number
+- String: A sequential collection of characters to represent a text
+- List: A list of values that can be access by index
+- Table: A dictionary to hold key and value pairs
+- Function: Function object which includes module level function, member
+  function or build-in function
+- Module: Module object which can hold functions and objects in it
+- SeedObject: The base type of all the customer objects
+
+Values of all types are the first-class values. They can be stored in global
+variables, local variables, object properties, list and table fields. They can
+be passed as arguments to functions, and return from functions as well.
+
+## Instruction Basics
+
+All SeedLang instructions are 32 bits fix-sized unsigned integer. There are
+three instruction types (iABC, iABx, iAsBx) and 24 (number from 0 to 23)
+opcodes. The layout of the three instruction types are illustrated as follows:
+
+```text
+         0 0 0 0 0 0 0 0 0 0 1 1 1 1 1 1 1 1 1 1 2 2 2 2 2 2 2 2 2 2 3 3
+         0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
+        -----------------------------------------------------------------
+iABC    | Opcode: 6 |     A: 8      |      B: 9       |      C: 9       |
+        -----------------------------------------------------------------
+iABx    | Opcode: 6 |     A: 8      |              Bx: 18               |
+        -----------------------------------------------------------------
+iAsBx   | Opcode: 6 |     A: 8      |             sBx: 18               |
+        -----------------------------------------------------------------
+```
+
+There are at most 64 opcodes for SeedLang byte code, because the opcode occupies
+the lower 6 bits of instruction.
+
+Instruction fields are encoded as simple unsigned integer values, except for
+sBx. Field sBx represents negative numbers, but it doesn't use 2s complement.
+Instead, it has a bias equal to half the maximum integer that can be represented
+by its unsigned counterpart, Bx. For a field size of 18 bits, Bx can hold a
+maximum unsigned integer value of 262143, and so the bias is 131071 (calculated
+as 262143 >> 1). A signed value of x will be encoded as (x + 131071).
+
+## Instruction Notation
+
+| Notation         | Description                                      |
+| ---------------- | ------------------------------------------------ |
+| PC               | Program Counter                                  |
```

Comment:
As I mentioned above, it will be way better to add the code highlighting notation to "PC", "R(A)", "A", "B", "C", etc. inside this table.